XBee settings sender/receiver

lhelwerd opened the pull request:
- Enable the control panel settings view to send settings to the vehicles (e1ba7e3).
- Reuse the waypoints sending code for the settings sending (4880ed3).
- Pack the settings key and value into the XBee packet using few bytes using a specialized string format and an object format that uses specific struct formats or JSON formatting and GZip compression (ce0c076, cleanup in later commits).
- Receive the settings in the vehicles to bundle them together and write it to a JSON settings file, also restarting the mission so that the settings are reloaded (b792481, wrapping up in later commits).
- Add methods to the settings form widgets to get the changed value, check if it is different from the default, and to reset them to the current override or the package default. The first two methods are used for the sending algorithm, and the reset functionality is exposed to context menus. Some other UX changes related to size policies were made (2fe9e9e).

The sender and receiver have been tested with the actual hardware; unit tests are also provided. This PR can best be examined commit-by-commit, and the UX tweaks might need some testing in the control panel to ensure that it works fine in other window managers as well.

lhelwerd committed:
XBee_Packet: Pack strings and other objects

Add a specification for the settings packet communications, which is to
be used by the Settings view save sender. The specification relies on
specially defined formats that allow us to pack a variable-length string
into the XBee packet. Additionally, we can send arbitrary JSON-encodable
objects that are compressed by Zlib. In the arbitrary object format, for
some basic types (bool, int, float and string), we use the smaller
packed format instead of the slower and bulkier compression steps.

This allows us to send settings in the packets without immediate concern
about packet sizes. Even the current value for "servo_pins", a list of
dictionaries, fits in the packet at less than 80 bytes in total.

lhelwerd committed:
Settings view: Reset via context menu and other UX

- Add actions to the context menus of the settings group boxes as well
  as the individual inputs. The two actions reset the value of the
  widget to either the current override or the settings default. This
  makes it easier to undo unintended changes when one detects them in
  the save dialog. Ensure that the text widget's line edit context menu
  has both the standard options and our actions, and do not give it the
  same background color as the validated input.
- When scrolling to a match after searching (or clearing the filter),
  also scroll the list widget so that the currently selected item is
  still in view.
- Change some size policies again. The parent button needs a vertical
  size policy to make it exactly as high as it needs to be. Otherwise,
  it will take up the same amount of space that the next form widget
  group box does, which is too much for small child components.
  A similar problem appears with the description text, which gets
  misaligned with its label when you add an item to a list form widget.

lhelwerd committed:
Waypoints view: Split out XBee sender methods

This allows us to reuse the sending progress tracker for other views.

lhelwerd committed:
Mission: Clean up waypoints to make more sense

- Add a takeoff command to the Mission_XBee and keep track of the
  existence of that command when returning the next index in the
  acknowledgement.
- Mock_Vehicle and parents: Count the waypoints from zero instead of
  a mix of one and zero, which occasionally causes the vehicle to skip
  or end prematurely. This did not happen in actual vehicles.
- Mission_Auto: Do not by default end when we are going to the last
  waypoint, since that immediately causes the vehicle to return to
  launch. Instead only do this in the Mission_Square, and keep a check
  for when we have done all waypoints.
- Clean up ancient documentation that was also in the wrong place.

lhelwerd committed:
Create XBee_Settings_Receiver, fix specifications

- Add a helper class that can be imported into any script that has an
  Environment to be able to receive settings via XBee packets. The XBee
  settings receiver knows the XBee, handles the specific packet
  specifications and tracks the settings coming in. Finally, it writes
  the settings and restarts when all settings are received. Its behavior
  is completely tested.
- Update the specifications for the setting packets to track the current
  numeric index as well as the setting key. This is necessary for the
  Control_Panel_XBee_Receiver which uses this for progress and tracking
  internally which data item to send. This adds 4 bytes to the packet,
  which should not be so devastating in most common cases.

lhelwerd committed:
Send settings from control panel to vehicles

Hook up the XBee sender to the confirmation dialog of the settings view.
If we want to send them to vehicles, start the sender that works in
a similar way to the waypoints sender. After it is done sending, save
them to the ground station settings (if we want to) as well.

Hook up the settings receiver to the Environment so that all missions
accept settings during the custom startup (non-measurement) phase.
Update the test to just use the created receiver.

Clean up the XBee_Packet serialization/unserialization error handling,
so that it becomes more obvious which specification and which field is
responsible for causing an error. Also makes the code a bit more
readable by moving blocks of code into their own method.

lhelwerd committed:
XBee settings: Clean up send/receive and interrupt

- Remove a debug print in the settings view packet sending method
- Call the thread manager's interrupt with the correct thread name
  instead of nothing (which causes an exception rather than directly
  interrupting all threads). This allows us to kill the threads as we
  are supposed to in the XBee settings receiver.
- Make the thread name of a Threadable object publicly readable, so that
  callback methods can grab them when they need to directly interrupt.
- Fix the test for the XBee settings receiver interrupt as well as add
  a test for the new thread name property.

timvandermeij commented on tests/xbee_packet.py:
```diff
@@ -104,6 +122,26 @@ def test_unserialize(self):
         })
         self.assertFalse(self.packet._private)
 
+    def test_unserialize_object_pack(self):
```
This should be `test_unserialize_object_packed(self)` (pack -> packed) to be in line with the serialization test name.

timvandermeij commented on tests/xbee_packet.py:
```diff
@@ -104,6 +122,26 @@ def test_unserialize(self):
 
+    def test_unserialize_object_pack(self):
+        self.packet.unserialize("\n\x03bar\x01i*\x00\x00\x00\x01")
+        self.assertEqual(self.packet._contents, {
```
We should use the `get_all` method here.

timvandermeij commented on tests/xbee_packet.py:
```diff
@@ -104,6 +122,26 @@ def test_unserialize(self):
+            "value": 42,
+            "to_id": 1
+        })
+        self.assertFalse(self.packet._private)
```
We should use the `is_private` method here.

timvandermeij commented on tests/xbee_packet.py:
```diff
@@ -104,6 +122,26 @@ def test_unserialize(self):
+
+    def test_unserialize_object_compressed(self):
+        self.packet.unserialize("\n\x05items\x00\x11x\x9c\x8b6\xd4Q0\xd2Q0\x8e\x05\x00\t\x85\x01\xe7\x01")
+        self.assertEqual(self.packet._contents, {
```
Same as above: `get_all`.

timvandermeij commented on tests/xbee_packet.py:
```diff
@@ -104,6 +122,26 @@ def test_unserialize(self):
+            "value": [1,2,3],
+            "to_id": 1
+        })
+        self.assertFalse(self.packet._private)
```
Same as above: `is_private`.

timvandermeij commented on zigbee/XBee_Packet.py:
```diff
@@ -140,8 +172,36 @@ def unserialize(self, contents):
+                is_packed, offset = self._read_packed("?", contents, offset)
+                if is_packed:
+                    object_format, offset = self._read_packed("B", contents,
+                                                                  offset)
```
The alignment is a bit off here. Let's align `offset)` with `"B"`.

lhelwerd committed:
Clean up XBee tests and layout

- Use XBee_Packet methods in all unserialize tests instead of accessing
  private member variables.
- Realign the arguments in some calls.

timvandermeij commented on control_panel/Control_Panel_Waypoints_View.py:
```diff
@@ -257,62 +251,15 @@ def _send(self):
-            return
-
-        waypoint = self._waypoints[vehicle][index]
+        sender = Control_Panel_XBee_Sender(self._controller, "waypoint",
```
Can we pass an object here instead of all these arguments? It helps keeping this readable.

timvandermeij commented on trajectory/Mission.py:
```diff
@@ -349,11 +349,9 @@ def check_waypoint(self):
-        num_commands = self.vehicle.count_waypoints()
-        if next_waypoint >= num_commands:
-            print("Exit 'standard' mission when heading for final waypoint ({})".format(num_commands))
+        if next_waypoint >= self.vehicle.count_waypoints():
```
Can we make this a oneliner? `return next_waypoint < self.vehicle.count_waypoints()`

timvandermeij commented on zigbee/XBee_Packet.py:
```diff
@@ -158,7 +167,7 @@ def unserialize(self, contents):
 
         if specification is None:
-            raise KeyError("Invalid specification has been provided")
+            raise KeyError("Invalid specification {} has been provided".format(specification_id))
```
Let's put the `{}` between single quotes for consistency.

lhelwerd committed:
Fix up XBee sending, and some nits

- Fix the accept callback so that it is actually called.
- Make a configuration object to pass to the sender.
- Fix XBee packet with string object format which needs to be packed
  recursively.
- Some other style cleanups from the PR.
